### PR TITLE
Extend pulse defaults

### LINKED ISSUE #2861: CmdDef should have a simpler representation
### What is the expected enhancement?
The current `CmdDef`'s, `repr` is based on its underlying dictionary. This quickly becomes complicated. Perhaps something like

`CmdDef(commands=['u1','u2','u3','measure','cx'], qubits=[0,1,2,3])`

is simpler to take in at a quick glance.

### LINKED ISSUE #2955: cmd_def.get error messaging
### What is the current behavior?
`CMdDef.get` error messaging returns a standard error. For example
```
cmd_def.get('u3', qubits=(1, 0), P0=2*np.pi, P1=0., P2=np.pi)
```
returns
```
'Command u3 for qubits (1, 0) is not present in CmdDef'
```
because `'u3'` is not a two-qubit gate (wrong user input).

### What is the expected enhancement?
A more informative error message could suggest that `'u3'` is present but that the `qubits` argument is wrong. Potentially, the error message could suggest the correct qubits format.

### PR DESCRIPTION
**Pull request by lcapelluto**
Discussion:
 - [ ] What about creating an empty defaults instance with no backend? (It can be helpful when writing tests to sometimes start with an empty `CmdDef`.)
 - [x] Think about the `op_qubits` name

#### Fixes:
Fixes #2955 (improved error message for CmdDef.get)
Fixes #2861 (improved repr and str for CmdDef)

#### Changes:
`qubit_freq_est` and `meas_freq_est` now return frequency in Hz, and they warn the user of this.

#### Replacing `CmdDef`
`CmdDef` will be deprecated following this PR. All existing methods are supported with the following exceptions:
- `__init__` is called by the provider, so it doesn't take a list of user defined operation schedules. `add` should be used instead.
- `cmds` and `cmd_qubits` are included with a deprecation message to use `ops` and `qubits_with_op` instead.

#### New methods (not in CmdDef):
- `qubit_ops(qubits)`: Return the operations that are defined for the qubit(s)
- `assert_has(operation, qubits)`: Raise a PulseError if the op isn't defined
- `remove(operation, qubits)`

## Details and comments

This should improve pulse usability. Rather than making more new objects to help users find the data they need, we extend the backend models with methods that will make it easier for the user to find the data they want.

This begins to separate the format required at the transport layer from the way we present the python objects on the Qiskit-side.

This PR only concerns `PulseDefaults`, found via `backend.defaults()`. I am allowing an accelerated migration (still fairly gentle) on this PR because `PulseDefaults` are only used by the Pulse system, which we consider to be in beta.

### CODE REVIEW

File: qiskit/providers/basebackend.py [hunk 1]
```diff
@@ -99,6 +99,14 @@ def name(self):
         """
         return self._configuration.backend_name
 
+    def version(self):
```
Comment (lcapelluto) [inline qiskit/providers/basebackend.py]: bonus method. could open a separate pr if desired.

File: qiskit/providers/models/pulsedefaults.py [hunk 2]
```diff
@@ -104,34 +110,326 @@ def __init__(self, name, **kwargs):
...
+            **kwargs (Dict[str, Any]]): Other attributes for the super class.
+        """
+        super().__init__(**kwargs)
+
         self.buffer = buffer
+        self._qubit_freq_est_ghz = qubit_freq_est
+        self._meas_freq_est_ghz = meas_freq_est
+        self._qubit_freq_est_hz = [freq * 1e9 for freq in qubit_freq_est]
+        self._meas_freq_est_hz = [freq * 1e9 for freq in meas_freq_est]
+        # TODO: These should be massaged for the user
```
Comment (lcapelluto) [inline qiskit/providers/models/pulsedefaults.py]: Since users touch this, I don't want to change it now

File: qiskit/schemas/examples/default_pulse_configuration_example.json [hunk 3]
```diff
@@ -24,7 +24,7 @@
         {"name": "measure", "qubits": [0],
         "sequence": [{"name": "square_1",
         "t0": 0, "ch": "m0"},
-        {"name": "acquire", "t0": 3, "duration": 10}]},
+        {"name": "acquire", "t0": 3, "duration": 10, "qubits":[0,1,2], "memory_slot": [0,1,2]}]},
```
Comment (lcapelluto) [inline qiskit/schemas/examples/default_pulse_configuration_example.json]: This was missing

File: qiskit/test/mock/fake_openpulse_2q.py [hunk 4]
```diff
@@ -89,10 +89,11 @@ def __init__(self):
                                        PulseQobjInstruction(name='test_pulse_1', ch='d1', t0=20),
                                        PulseQobjInstruction(name='fc', ch='d1',
                                                             t0=20, phase=2.1)]),
-                     Command(name='measure', qubits=[0],
+                     Command(name='measure', qubits=[0, 1],
```
Comment (lcapelluto) [inline qiskit/test/mock/fake_openpulse_2q.py]: Update to match what `meas_map` claimed

File: qiskit/providers/models/pulsedefaults.py [hunk 5]
```diff
@@ -104,34 +110,326 @@ def __init__(self, name, **kwargs):
...
+        Return the estimated measurement stimulus frequency to readout from the given qubit.
+
+        Returns:
+            list[float]: The measurement stimulus frequency in Hz.
+        """
+        warnings.warn("The measurement frequency estimation was previously in GHz, "
+                      "and now is in Hz.")
+        return self._meas_freq_est_hz
+
+    def replace_pulse(self, name, samples):
```
Comment (dcmckayibm) [inline qiskit/providers/models/pulsedefaults.py]: Does this deal with refactoring the schedule so that the alignments are the same? Eg. if we use this to replace the pi pulse in our definitions for the CNOT will it move the CR pulses appropriately?

Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: Ah, it was supposed to, but I'm realizing now that the way it's written will not work, because the default schedules specify exact times to play pulses. I'll have to fix that. Thank you!! 🙏

Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: Moving this to a separate PR

File: qiskit/providers/models/pulsedefaults.py [hunk 6]
```diff
@@ -94,44 +101,325 @@ class Command(BaseModel):
...
+            The names of all the circuit operations which have Schedule definitions in this.
+        """
+        return list(self._ops_def.keys())
 
-    def build_cmd_def(self):
-        """Construct the `CmdDef` object for the backend.
+    def op_qubits(self, operation: str) -> List[Union[int, Tuple[int]]]:
+        """
+        Return a list of the qubits for which the given operation is defined. Single qubit
+        operations return a flat list, and multiqubit operations return a list of ordered tuples.
```
Comment (eggerdj) [inline qiskit/providers/models/pulsedefaults.py]: `op_qubits` is almost too short of a name. Can we have a name that has more meaning? The first thing that comes to mind is `qubits_with_operation` but that almost seems too wordy (or maybe `qubits_with_op`). Suggestions?
```
cx_qubits = qubits_with_op('cx')
```

File: qiskit/providers/models/pulsedefaults.py [hunk 7]
```diff
@@ -94,44 +101,325 @@ class Command(BaseModel):
...
+
+        Args:
+            operation: The operation for which to look.
+            qubits: The specific qubits for the operation.
+
+        Returns:
+            True iff the operation is defined.
+        """
+        return operation in self._ops_def and \
+            _to_tuple(qubits) in self._ops_def[operation]
```
Comment (eggerdj) [inline qiskit/providers/models/pulsedefaults.py]: Can we clarify in the doc string what the `qubits` input should look like for single-qubit gates versus multi-qubit gates?

File: qiskit/providers/models/pulsedefaults.py [hunk 8]
```diff
@@ -94,44 +101,325 @@ class Command(BaseModel):
...
+
+    def assert_has(self, operation: str, qubits: Union[int, Iterable[int]]) -> None:
+        """
+        Convenience method to check that the given operation is defined, and error if it is not.
+
```
Comment (eggerdj) [inline qiskit/providers/models/pulsedefaults.py]: Same comment as above regarding `qubits` input.

File: qiskit/providers/models/pulsedefaults.py [hunk 9]
```diff
@@ -46,17 +53,17 @@ class CommandSchema(BaseSchema):
     name = String(required=True)
 
     # Optional properties.
-    qubits = List(Integer(validate=Range(min=0)),
-                  validate=Length(min=1))
+    qubits = QList(Integer(validate=Range(min=0)),
+                   validate=Length(min=1))
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: yes, but we decided we didn't like this as much as `fields.List`, so I'm going to update it with the improved conflict resolution

File: qiskit/providers/models/pulsedefaults.py [hunk 10]
```diff
@@ -94,44 +101,325 @@ class Command(BaseModel):
...
+    @property
+    def qubit_freq_est(self) -> List[float]:
+        """
+        Return the estimated resonant frequency for the given qubit in Hz.
+
+        Returns:
+            The frequency of the qubit resonance in Hz.
+        """
+        warnings.warn("The qubit frequency estimation was previously in GHz, and now is in Hz.")
+        return self._qubit_freq_est_hz
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: Up to 2 months (maybe less if we release more frequently). Open to suggestions

File: qiskit/providers/models/pulsedefaults.py (hunk 6 above)
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: I was trying to be a bit consistent with `cmd_qubits`. I will definitely put some thought into this.

File: qiskit/providers/models/pulsedefaults.py (hunk 7 above)
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: Is it not clear from the type hints? There is nothing distinguishing the input for single vs multi qubit gates (except that inherently, it will be difficult to get a multi qubit gate without passing multiple qubits!)

File: qiskit/test/mock/fake_openpulse_2q.py (hunk 4 above)
Comment (lcapelluto) [reply qiskit/test/mock/fake_openpulse_2q.py]: Yes, the above works as it is written. (Did I misunderstand the question? Why would you need two measure commands?)

File: qiskit/util.py [hunk 11]
```diff
@@ -95,3 +95,18 @@ def _has_connection(hostname, port):
...
+
+    Args:
+        values (Union[int, Iterable[int]]): An integer, or iterable of integers.
+    Returns:
+        tuple: The input values as a sorted tuple.
+    """
+    try:
+        return tuple(values)
+    except TypeError:
+        return (values,)
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/util.py]: `cx` on 0 1 is not the same gate as `cx` on 1 0, if you look in the `cmd_def` returned by some backends.

File: qiskit/compiler/assemble.py [hunk 12]
```diff
@@ -247,8 +248,12 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
     schedule_los = [lo_config if isinstance(lo_config, LoConfig) else LoConfig(lo_config)
                     for lo_config in schedule_los]
 
-    qubit_lo_freq = qubit_lo_freq or getattr(backend_default, 'qubit_freq_est', None)
-    meas_lo_freq = meas_lo_freq or getattr(backend_default, 'meas_freq_est', None)
+    warnings.filterwarnings("ignore")
+    if not qubit_lo_freq and hasattr(backend_default, 'qubit_freq_est'):
+        qubit_lo_freq = [freq / 1e9 for freq in backend_default.qubit_freq_est]
+    if not meas_lo_freq and hasattr(backend_default, 'meas_freq_est'):
+        meas_lo_freq = [freq / 1e9 for freq in backend_default.meas_freq_est]
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/compiler/assemble.py]: Yes. It's confusing and unfortunate. I think a good follow up would be to keep Hz throughout. The conversion to GHz would still have to take place when inserting the values into the `Qobj` level, though. It's unfortunate that that prefix is baked into the spec

File: qiskit/providers/models/pulsedefaults.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: One alternative idea would be to create a `qubit_freq_est_hz` and then raise a deprecation warning for the GHz `qubit_freq_est` (but would want `qubit_freq_est` to eventually be Hz). Since Pulse is in beta, I thought it would be appropriate to use a quicker deprecation strategy

File: test/python/compiler/test_assembler.py [hunk 13]
```diff
@@ -533,8 +533,8 @@ def setUp(self):
         self.backend = FakeOpenPulse2Q()
         self.config = self.backend.configuration()
         self.defaults = self.backend.defaults()
-        self.qubit_lo_freq = self.defaults.qubit_freq_est
-        self.meas_lo_freq = self.defaults.meas_freq_est
+        self.qubit_lo_freq = [freq / 1e9 for freq in self.defaults.qubit_freq_est]
+        self.meas_lo_freq = [freq / 1e9 for freq in self.defaults.meas_freq_est]
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply test/python/compiler/test_assembler.py]: yes

File: qiskit/compiler/assemble.py [hunk 14]
```diff
@@ -247,8 +248,12 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
     schedule_los = [lo_config if isinstance(lo_config, LoConfig) else LoConfig(lo_config)
                     for lo_config in schedule_los]
 
-    qubit_lo_freq = qubit_lo_freq or getattr(backend_default, 'qubit_freq_est', None)
-    meas_lo_freq = meas_lo_freq or getattr(backend_default, 'meas_freq_est', None)
+    warnings.filterwarnings("ignore")
```
Comment (1ucian0) [inline qiskit/compiler/assemble.py]: why?

File: qiskit/providers/models/pulsedefaults.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: @1ucian0

File: qiskit/compiler/assemble.py (hunk 14 above)
Comment (lcapelluto) [reply qiskit/compiler/assemble.py]: yeah, the `qubit_freq_est` raises a warning because its behavior has changed. I don't want this warning showing when users are using the assembler, since it isn't their concern (the assembler is working in sync with the pulse defaults changes). I mentioned you in another thread -- I would be interested in discussing alternative migration paths for `qubit_freq_est` and `meas_freq_est`

File: qiskit/providers/models/pulsedefaults.py (hunk 6 above)
Comment (lcapelluto) [reply qiskit/providers/models/pulsedefaults.py]: I took the suggestion, thank you @eggerdj

**Comment by lcapelluto** [comment]
There is some overlap with another merged branch. This branch would revert those changes. Going to close this PR and open a clean one.
